# UT6: EDA - Estadística Descriptiva Aplicada

Análisis de datos con Python

# UT6: Análisis Exploratorio de Datos (EDA)

> **Cuaderno de trabajo — UT6: Análisis Exploratorio de Datos (EDA)**
>
> Este notebook contiene los bloques de código y ejercicios de la
> unidad. Para la teoría completa consulta el libro (PDF).
>
> **Requisitos:** ejecuta primero la celda de preparación del entorno y
> la de carga del *dataset*.

## El rol del EDA en el ciclo de vida del proyecto

**EDA inicial vs. EDA profundo:** no son lo mismo.

-   **EDA inicial:** se realiza *antes* de la limpieza, para
    diagnosticar qué problemas tiene el *dataset*.
-   **EDA profundo:** se realiza *después* de la limpieza (es el que
    aborda esta UT), para descubrir relaciones, patrones y supuestos que
    guiarán el modelado.

Confundirlos es uno de los errores más frecuentes en proyectos reales.

Un compañero ejecuta `df.isnull().sum()` sobre el *dataset* crudo y
declara «ya hice el EDA». ¿Qué fase del EDA ha realizado realmente? ¿Qué
información crítica le falta para tomar decisiones de modelado?

## Las 4 preguntas clave del EDA

## Caso de uso: predicción de precios de viviendas

Trabajamos con el dataset **Ames Housing** que ya limpiaste en UT5:
1.458 viviendas, 74 variables limpias (sin nulos ni *outliers*
extremos). El EDA de esta unidad determinará cuáles de esas 74 variables
merecen entrar en el modelo. Si necesitas refrescar el contexto del
dataset, consulta la nota introductoria en UT5.

Con ese objetivo claro, comenzamos preparando el entorno de trabajo.

## Preparación del entorno

In [1]:
# Librerías necesarias para esta unidad
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats

# Configuración de visualización
plt.style.use('default')            # estilo limpio sin fondo gris
sns.set_palette("husl")             # paleta perceptualmente uniforme
pd.set_option('display.max_columns', None)  # mostrar todas las columnas
pd.set_option('display.precision', 2)       # 2 decimales en tablas

print(f"Pandas: {pd.__version__}")
print(f"NumPy: {np.__version__}")
print("Entorno configurado correctamente")

Con el entorno listo, cargamos el *dataset* sobre el que trabajaremos a
lo largo de toda la unidad.

## Carga del *dataset*

In [2]:
BASE = 'https://raw.githubusercontent.com/jgarcia314/'
url = BASE + 'analisis-datos-python-fp/main/data/raw/house_prices_clean.csv'
df = pd.read_csv(url)
print(f"Dataset cargado: {df.shape[0]} filas × {df.shape[1]} columnas")
print(f"Variables numéricas: {df.select_dtypes(include='number').shape[1]} | Categóricas: {df.select_dtypes(include='object').shape[1]}")

Dataset cargado: 1458 filas × 74 columnas
Variables numéricas: 37 | Categóricas: 37

**Documentación del EDA**

En proyectos profesionales, siempre documenta tu EDA en un Jupyter
Notebook separado. Esto permite que otros (y tú mismo en el futuro)
entiendan las decisiones que tomaste basándote en los datos.

Con el *dataset* cargado y documentado, comenzamos el análisis variable
por variable.

## Análisis univariante: entendiendo variable por variable

### Primera inspección del *dataset*

In [3]:
df.shape        # filas × columnas
df.head()       # primeras filas
df.info()       # tipos y nulos
df.dtypes.value_counts()  # resumen de tipos

### Variables numéricas: más allá de la media

### Medidas de tendencia central revisitadas

``` python
media = df['SalePrice'].mean()
mediana = df['SalePrice'].median()

resumen = pd.DataFrame({
    'Medida': ['Media', 'Mediana'],
    'Valor (USD)': [f"{media:,.0f}", f"{mediana:,.0f}"],
})
print(resumen.to_markdown(index=False))
print("\nTable: Medidas de tendencia central de SalePrice — Ames Housing")
```

| Medida  | Valor (USD) |
|:--------|------------:|
| Media   |     180,933 |
| Mediana |     163,000 |

Medidas de tendencia central de SalePrice — Ames Housing

**Análisis de los resultados:** La media (\$180,933) supera a la mediana
(\$163,000) en \$17,933 (un 11.0 %), señal clara de asimetría positiva:
existe una cola de viviendas de lujo que eleva el promedio. En un
contexto de negocio inmobiliario, esto implica que la mediana es el
indicador de precio más representativo del mercado típico, y que
cualquier modelo entrenado con la media como referencia tendería a
sobrestimar precios en el segmento mayoritario.

La media de `SalePrice` (\$180,933) supera a la mediana (\$163,000) en
un 11 %. ¿Por qué sería incorrecto usar la media como precio de
referencia en una campaña de marketing inmobiliario dirigida al segmento
mayoritario del mercado de Ames? ¿Qué indicador publicarías en su lugar?

**Recuerda (UT5):** Los *outliers* distorsionan la media de forma
significativa pero apenas afectan a la mediana — por eso la mediana es
la medida de centro recomendada en distribuciones asimétricas o con
valores extremos. Lo viste en detalle en la UT5, en la sección de
detección de *outliers*.

### Medidas de dispersión profundas

| Medida                            | Qué mide                                            | Cuándo usarla                                                 |
|:-----------------|:---------------------|:-------------------------------|
| **Rango**                         | Amplitud total de los datos                         | Visión rápida, muy sensible a *outliers*                      |
| **Varianza**                      | Dispersión promedio al cuadrado respecto a la media | Base para otros cálculos, difícil de interpretar directamente |
| **Desviación estándar**           | Dispersión promedio en unidades originales          | La más usada: mismas unidades que los datos                   |
| **IQR**                           | Amplitud del 50% central de los datos               | Robusta ante *outliers*; ya la usaste en UT5 para detectarlos |
| **Coeficiente de Variación (CV)** | Dispersión relativa como porcentaje de la media     | Comparar variabilidad entre variables con escalas distintas   |

Medidas de dispersión — el CV es el concepto nuevo respecto a UT2 y UT5.

Apliquemos estas medidas sobre `SalePrice` para ver qué nos dicen en la
práctica.

``` python
precio = df['SalePrice']

rango = precio.max() - precio.min()
std = precio.std()
cv = (std / precio.mean()) * 100
q1, q3 = precio.quantile(0.25), precio.quantile(0.75)
iqr = q3 - q1

dispersion = pd.DataFrame({
    'Medida': ['Mínimo', 'Máximo', 'Rango', 'Desv. estándar', 'CV (%)', 'Q1 (25%)', 'Q3 (75%)', 'IQR'],
    'Valor': [
        f"${precio.min():,.0f}", f"${precio.max():,.0f}", f"${rango:,.0f}",
        f"${std:,.0f}", f"{cv:.2f}%",
        f"${q1:,.0f}", f"${q3:,.0f}", f"${iqr:,.0f}"
    ]
})
print(dispersion.to_markdown(index=False))
print("\nTable: Medidas de dispersión de SalePrice — Ames Housing")
```

| Medida         | Valor     |
|:---------------|:----------|
| Mínimo         | \$34,900  |
| Máximo         | \$755,000 |
| Rango          | \$720,100 |
| Desv. estándar | \$79,495  |
| CV (%)         | 43.94%    |
| Q1 (25%)       | \$129,925 |
| Q3 (75%)       | \$214,000 |
| IQR            | \$84,075  |

Medidas de dispersión de SalePrice — Ames Housing

**Análisis de los resultados:** El rango bruto (\$720.100, desde
\$34.900 hasta \$755.000) confirma la presencia de viviendas muy
atípicas que distorsionarán cualquier análisis basado en la media. La
desviación estándar de \$79.443 y el CV del 43,91 % indican alta
heterogeneidad: los precios no son predecibles por la media. El IQR de
\$84.025 (entre \$129.975 y \$214.000) es la medida más fiable: el 50 %
central de las viviendas del *dataset* oscila dentro de esa banda, que
representa el mercado residencial estándar de Ames. Los valores fuera de
ese rango merecen atención especial en el análisis de *outliers*.

El IQR es una medida de dispersión robusta que no se deja distorsionar
por valores extremos. En UT5 lo usamos para *detectar* outliers; aquí lo
usamos para *cuantificar* la variabilidad central del dataset.

El IQR de `SalePrice` es \$84,075 (entre \$129,925 y \$214,000). Un
analista filtra todas las viviendas fuera de ese rango antes de entrenar
el modelo. ¿Qué consecuencia tendría esa decisión para clientes que
buscan viviendas de lujo (\> \$400.000)? ¿Cuándo está justificado
filtrar por IQR antes de modelar?

``` python
variables_numericas = ['SalePrice', 'GrLivArea', 'YearBuilt', 'OverallQual']

cv_comparacion = (
    df[variables_numericas]
    .agg(lambda col: col.std() / col.mean() * 100)
    .rename("CV (%)")
    .round(2)
    .to_frame()
    .sort_values("CV (%)", ascending=False)
)
print(cv_comparacion.to_markdown())
print("\nTable: Coeficiente de variación por variable — ordenado de mayor a menor heterogeneidad")
```

|             | CV (%) |
|:------------|-------:|
| SalePrice   |  43.94 |
| GrLivArea   |  33.62 |
| OverallQual |  22.59 |
| YearBuilt   |   1.53 |

Coeficiente de variación por variable — ordenado de mayor a menor
heterogeneidad

**Análisis de los resultados:** El CV revela diferencias dramáticas
entre variables: `SalePrice` (43.94 %) y `GrLivArea` (33.62 %) son
altamente heterogéneas — el precio y el área habitable varían mucho
entre viviendas —, mientras que `YearBuilt` (1.53 %) es casi constante
en comparación, pues todas las casas se construyeron en rangos
temporales similares. `OverallQual` (22.59 %) muestra variabilidad
moderada. Esta comparación justifica por qué el CV es imprescindible:
sin él, compararíamos desviaciones en dólares, metros cuadrados y años,
magnitudes incomparables entre sí.

El CV de `SalePrice` (43.94 %) es muy superior al de `YearBuilt` (1.53
%). ¿Qué nos dice esa diferencia sobre qué variable será más difícil de
predecir con una media simple? ¿Cómo afecta un CV alto a la fiabilidad
de un modelo que usa la media como predictor base?

**Manos a la Obra**

#### Comparación de medidas de dispersión

**Objetivo:** Entender profundamente cuándo y por qué usar cada medida
de dispersión analizando variables con diferentes escalas y unidades.

**Contexto profesional:** Como analista de riesgos en una consultora
inmobiliaria, debes reportar qué características de las viviendas
presentan mayor volatilidad.

**Instrucciones:**

1.  Analiza estas tres variables: `LotArea` (área del terreno),
    `YearBuilt` (año de construcción) y `GarageArea` (área del garaje).
2.  Para cada una, calcula: Desviación estándar, Rango Intercuartílico
    (IQR) y Coeficiente de Variación (CV).
3.  **Análisis de datos:** Compara el CV de `LotArea` frente al de
    `GarageArea`. ¿Cuál de las dos características es más “impredecible”
    en relación a su magnitud media? Demuestra tu conclusión con los
    números obtenidos.
4.  Explica por qué el CV es la única medida justa para comparar la
    dispersión de `YearBuilt` (escala temporal) con `LotArea` (escala
    espacial).

**Criterio de éxito:** El código debe mostrar una tabla comparativa con
las 3 métricas para las 3 variables. La respuesta a la pregunta 3 debe
estar fundamentada en el cálculo del CV.

In [14]:
# Escribe tu codigo aqui


**Advertencia:** El CV solo tiene sentido para variables con **cero
absoluto** (donde cero = ausencia total). No lo uses cuando el cero es
arbitrario.

-   **Válido:** peso (0 kg = sin peso), altura, ingresos (0€ = sin
    ingresos)
-   **No válido:** temperatura Celsius (0°C es el punto de congelación,
    no “sin temperatura”), años

Las medidas de dispersión nos dicen cuánto varían los datos, pero no
cómo se distribuyen. Para eso necesitamos estudiar la forma de la
distribución.

### Forma de la distribución: asimetría y curtosis

| Tipo                                | Valor                          | Forma                     | Interpretación                                                       |
|:------------|:--------------|:--------------|:------------------------------|
| **Simétrica**                       | $\approx$ 0 (entre -0.5 y 0.5) | Media $\approx$ Mediana   | Datos equilibrados a ambos lados                                     |
| **Asimétrica positiva** (derecha)   | \> 0.5                         | Cola larga a la derecha   | Mayoría de valores bajos, algunos muy altos (ej: salarios, precios)  |
| **Asimétrica negativa** (izquierda) | \< -0.5                        | Cola larga a la izquierda | Mayoría de valores altos, algunos muy bajos (ej: edad de jubilación) |

Asimetría/Skewness

| Tipo             | Valor                      | Forma                    | Interpretación                                      |
|:------------|:--------------|:--------------|:------------------------------|
| **Mesocúrtica**  | $\approx$ 0 (entre -1 y 1) | Similar a la normal      | Comportamiento «típico»                             |
| **Leptocúrtica** | \> 1                       | Pico alto, colas pesadas | Más concentración en el centro + más *outliers*     |
| **Platicúrtica** | \< -1                      | Más plana, colas ligeras | Datos más uniformemente dispersos, menos *outliers* |

Curtosis

En Pandas, `.kurt()` devuelve la curtosis relativa a la normal: valor 0
= comportamiento normal, positivo = colas más pesadas, negativo = colas
más ligeras.

``` python
precio = df['SalePrice']
skewness = precio.skew()
kurtosis = precio.kurt()

forma = pd.DataFrame({
    'Medida': ['Asimetría (skewness)', 'Curtosis (kurtosis)'],
    'Valor': [round(skewness, 2), round(kurtosis, 2)],
    'Interpretación': [
        'Asimétrica a la derecha — cola de viviendas de lujo' if skewness > 0.5
        else ('Asimétrica a la izquierda' if skewness < -0.5 else 'Aproximadamente simétrica'),
        'Leptocúrtica — pico alto, colas pesadas (más outliers)' if kurtosis > 1
        else ('Platicúrtica — distribución más plana' if kurtosis < -1 else 'Mesocúrtica — similar a normal')
    ]
})
print(forma.to_markdown(index=False))
print("\nTable: Forma de la distribución de SalePrice — asimetría y curtosis")
```

| Medida               | Valor | Interpretación                                         |
|:------------------|--------:|:--------------------------------------------|
| Asimetría (skewness) |  1.88 | Asimétrica a la derecha — cola de viviendas de lujo    |
| Curtosis (kurtosis)  |  6.52 | Leptocúrtica — pico alto, colas pesadas (más outliers) |

Forma de la distribución de SalePrice — asimetría y curtosis

**Análisis de los resultados:** Con una asimetría (*skewness*) de 1.881,
`SalePrice` presenta asimetría positiva fuerte: la cola de viviendas
caras se extiende hacia la derecha, alejando la media de la mediana. La
curtosis de 6.523 (leptocúrtica) indica colas más pesadas de lo normal,
es decir, hay más *outliers* extremos de los esperados en una
distribución gaussiana. Ambos valores superan los umbrales críticos
(\|skew\| \> 1 y kurtosis \> 3), lo que confirma que una transformación
logarítmica es necesaria antes de aplicar modelos que asumen normalidad
como la regresión lineal. Tras aplicar $\log$, el *skewness* cae a 0.122
y la curtosis a 0.805 — prácticamente normales.

`SalePrice` tiene *skewness* = 1.88 y *kurtosis* = 6.52. ¿Qué dos
problemas concretos generan estos valores en un modelo de regresión
lineal que asume normalidad? ¿La transformación logarítmica resuelve
ambos o solo el primero?

``` python
precio_original = df['SalePrice']
precio_log = np.log(df['SalePrice'])

comparacion = pd.DataFrame({
    'Variable': ['SalePrice (original)', 'log(SalePrice)'],
    'Skewness': [round(precio_original.skew(), 3), round(precio_log.skew(), 3)],
    'Interpretación': [
        'Asimetría positiva fuerte' if precio_original.skew() > 1 else 'Moderada',
        'Prácticamente simétrica' if abs(precio_log.skew()) < 0.5 else 'Moderada'
    ]
})
print(comparacion.to_markdown(index=False))
print("\nTable: Impacto de la transformación logarítmica sobre la asimetría de SalePrice")
```

| Variable             | Skewness | Interpretación            |
|:---------------------|---------:|:--------------------------|
| SalePrice (original) |    1.881 | Asimetría positiva fuerte |
| log(SalePrice)       |    0.122 | Prácticamente simétrica   |

Impacto de la transformación logarítmica sobre la asimetría de SalePrice

Una asimetría con valor absoluto superior a 1 es una señal para
investigar si una transformación logarítmica puede mejorar el modelo —
pero no una regla automática. Dos condiciones previas: la variable debe
ser estrictamente positiva (log no está definido en cero o negativos) y
hay que verificar empíricamente que la transformación mejora realmente
el ajuste.

**Manos a la Obra**

#### Transformaciones y su impacto

**Objetivo:** comprender cuándo y por qué transformar variables
asimétricas.

**Contexto profesional:** Estás preparando datos para un modelo de
regresión lineal que asume normalidad. Debes identificar qué variables
requieren transformación y justificar tu decisión.

**Instrucciones (Semi-guiado):**

Analiza las variables `SalePrice`, `GrLivArea` y `LotArea`:

1.  Para cada una: calcula *skewness* y *kurtosis*, e identifica si es
    simétrica o asimétrica.
2.  Si es asimétrica (`|skew| > 0.5`), aplica transformación logarítmica
    y compara *skewness* antes/después. Verifica visualmente la mejora.
3.  Comprueba qué ocurre si aplicas log a `OverallQual`, que ya es
    aproximadamente simétrica: ¿mejora o empeora el *skewness*?
4.  Si una variable tiene valores de 0, no puedes aplicar log
    directamente. ¿Qué alternativa usarías?

El *skewness* de `SalePrice` pasó de 1,8 a 0,3 tras la transformación
logarítmica. ¿Qué significa eso en términos prácticos para un modelo de
regresión lineal? ¿Por qué la transformación log reduce la asimetría en
distribuciones lognormales?

Transformaste `SalePrice` con `np.log1p()` para entrenar el modelo. Este
predice en escala logarítmica. ¿Qué transformación inversa debes aplicar
antes de presentar los resultados al equipo de negocio? ¿Qué error se
comete si lo olvidas?

**Criterio de éxito:**

-   Has identificado correctamente qué variables son asimétricas.
-   Has aplicado la transformación solo donde es necesaria.
-   Puedes explicar el impacto de la transformación en números y
    palabras.

Consulta el **Apéndice C.2** para ver cómo calcular la asimetría y
curtosis en Pandas.

``` python
# Escribe aquí tu código
```

No transformes automáticamente todas las variables asimétricas. Pregunta
primero: ¿mi modelo requiere normalidad? Árboles de decisión y modelos
no paramétricos no necesitan transformaciones.

Tienes dos variables asimétricas: `SalePrice` (*skew* = 1,88) para una
regresión lineal, y `OverallQual` (*skew* = −0,12) para un Random
Forest. ¿A cuál de las dos aplicarías transformación logarítmica?
¿Cambiaría tu respuesta si ambas entraran en el mismo Random Forest?

En proyectos reales con decenas de variables numéricas, calcular estas
métricas columna por columna es inviable. El paso natural es
encapsularlas en una función reutilizable — y el hábito de construir tu
propia biblioteca de funciones de análisis es una de las mejores
inversiones que puedes hacer como analista. En el apéndice (Cheat Sheet)
encontrarás una implementación de referencia lista para usar.

### Variables categóricas: frecuencias y proporciones

El primer paso es siempre construir una tabla de frecuencias que muestre
no solo cuántas veces aparece cada categoría, sino qué peso relativo
tiene en el conjunto.

### Tablas de frecuencia enriquecidas

``` python
frecuencias = df['Neighborhood'].value_counts()
proporciones = df['Neighborhood'].value_counts(normalize=True)

tabla = pd.DataFrame({
    'Frecuencia': frecuencias,
    'Porcentaje': (proporciones * 100).round(1),
    'Acumulado %': (proporciones * 100).cumsum().round(1)
})
print(tabla.head(10).to_markdown())
print("\nTable: Top 10 barrios por frecuencia — Neighborhood en Ames Housing")
```

| Neighborhood | Frecuencia | Porcentaje | Acumulado % |
|:-------------|-----------:|-----------:|------------:|
| NAmes        |        225 |       15.4 |        15.4 |
| CollgCr      |        150 |       10.3 |        25.7 |
| OldTown      |        113 |        7.8 |        33.5 |
| Edwards      |         98 |        6.7 |        40.2 |
| Somerst      |         86 |        5.9 |        46.1 |
| Gilbert      |         79 |        5.4 |        51.5 |
| NridgHt      |         77 |        5.3 |        56.8 |
| Sawyer       |         74 |        5.1 |        61.9 |
| NWAmes       |         73 |          5 |        66.9 |
| SawyerW      |         59 |          4 |        70.9 |

Top 10 barrios por frecuencia — Neighborhood en Ames Housing

**Análisis de los resultados:** `Neighborhood` tiene 25 categorías. Las
10 primeras por frecuencia acumulan aproximadamente el 71 % de los
registros: NAmes lidera con 225 viviendas (15,4 %), seguido de CollgCr
(150), OldTown (113) y Edwards (100). En el extremo opuesto, Blueste
solo aparece en 2 registros (0,1 %). Esta distribución desigual es
típica en variables geográficas: unos pocos barrios concentran la
actividad del mercado y los demás tienen representación anecdótica.

`Blueste` solo tiene 2 registros en el *dataset* (0,1 %). Si la incluyes
como categoría independiente en el modelo, ¿qué tipo de error estás
arriesgando? ¿Es siempre mejor agrupar categorías minoritarias bajo
«Otros» o hay casos en los que conviene conservarlas separadas?

### Concentración de Categorías

``` python
acumulado_cum = (frecuencias / frecuencias.sum()).cumsum()
n_80 = (acumulado_cum < 0.80).sum() + 1
pct_concentracion = n_80 / len(frecuencias) * 100
interpretacion = (
    "Altamente concentrado — pocas categorías dominan" if pct_concentracion < 20
    else ("Moderadamente concentrado" if pct_concentracion < 50
          else "Distribuido uniformemente")
)

resumen = pd.DataFrame({
    'Métrica': ['Total de categorías', 'Categorías que acumulan el 80 %',
                'Nivel de concentración', 'Interpretación'],
    'Valor': [len(frecuencias), n_80,
              f'{pct_concentracion:.1f} %', interpretacion]
})
print(resumen.to_markdown(index=False))
print("\nTable: Análisis de concentración de Neighborhood — Ames Housing")
```

| Métrica                         | Valor                     |
|:--------------------------------|:--------------------------|
| Total de categorías             | 25                        |
| Categorías que acumulan el 80 % | 13                        |
| Nivel de concentración          | 52.0 %                    |
| Interpretación                  | Distribuido uniformemente |

Análisis de concentración de Neighborhood — Ames Housing

**Análisis de los resultados:** Para `Neighborhood`, 13 de las 25
categorías concentran el 80 % de los registros, lo que da un nivel de
concentración del 52 %: moderado según la escala del código. Eso
significa que las 12 categorías restantes comparten el 20 % de los datos
entre ellas; varias de ellas apenas superan los 10 registros, umbral
mínimo para que una categoría sea estadísticamente fiable. La señal
práctica es clara: si entrenas un modelo con las 25 categorías tal cual,
las 14 minoritarias añadirán ruido. Agrupa las que tengan menos del 1 %
en una categoría “Otros” y gana estabilidad sin sacrificar información
relevante.

El análisis de Pareto muestra que 13 de los 25 barrios concentran el 80
% de los registros. ¿Por qué esa desigualdad puede perjudicar a un
modelo de clasificación? ¿Qué estrategia usarías para que el modelo no
ignore los barrios con menor representación?

**Manos a la Obra**

#### Análisis de concentración categórica

**Objetivo:** Identificar categorías significativas y decidir
estrategias de agrupación basadas en la frecuencia y la
representatividad.

**Contexto profesional:** Debes optimizar el pre-procesamiento de
variables categóricas para evitar el “ruido” estadístico que generan las
clases con muy pocos ejemplos.

**Instrucciones (Semi-guiado):**

1.  Analiza las variables `BldgType` (tipo de edificio) y `Condition1`
    (proximidad a vías principales).
2.  Para ambas, calcula la frecuencia absoluta y el porcentaje que
    representa cada categoría sobre el total.
3.  Identifica cuántas categorías concentran el **80% de los datos**
    (análisis de Pareto).
4.  Localiza aquellas categorías que aparezcan en **menos del 1%** del
    dataset.
5.  **Análisis por datos:** Calcula el precio medio (`SalePrice`) para
    las categorías mayoritarias frente a las minoritarias de `BldgType`.
    ¿La diferencia de precio medio supera el 15 %? Si es así, hay
    razones para mantenerlas separadas aunque sean poco frecuentes; si
    no, agrúpalas como “Otros”.

**Pistas:**

-   Usa `.value_counts(normalize=True)` para obtener proporciones.
-   Usa `.groupby().agg()` para cruzar con el precio de venta.

**Criterio de éxito:** El código debe identificar las categorías que
concentran el 80% de la muestra y listar aquellas que deberían ser
agrupadas como ‘Otros’ por su baja frecuencia.

In [26]:
# Escribe tu codigo aqui


No agrupes categorías mecánicamente por frecuencia. Primero verifica si
tienen comportamientos similares respecto al target (precio). Agrupa
solo categorías que sean conceptualmente similares Y tengan valores de
target comparables.

Del mismo modo, cuando trabajes con decenas de variables categóricas,
encapsula el análisis de frecuencias, concentración y categorías raras
en una función reutilizable. En el apéndice (Cheat Sheet) encontrarás la
implementación de referencia junto a la de variables numéricas.

### Consolidación: análisis univariante

## Análisis bivariante: relaciones entre variables

### Numérica vs. numérica: correlaciones y patrones

### Correlación de Pearson en Profundidad

**Nomenclatura:** Este concepto se conoce por varios nombres
equivalentes:

-   **Coeficiente de correlación** (forma abreviada común)
-   **Coeficiente de correlación de Pearson**
-   **Coeficiente de correlación lineal de Pearson**
-   **r de Pearson** (en notación estadística se representa como *r*)
    $$r = \frac{\sum_{i=1}^{n} (x_i - \bar{x})(y_i - \bar{y})}{\sqrt{\sum_{i=1}^{n} (x_i - \bar{x})^2 \sum_{i=1}^{n} (y_i - \bar{y})^2}}$$

Todos se refieren a lo mismo: una medida entre -1 y +1 que cuantifica la
relación lineal entre dos variables. Fue desarrollado por Karl Pearson a
finales del siglo XIX.

``` python
correlacion = df['GrLivArea'].corr(df['SalePrice'])
r_cuadrado = correlacion ** 2
fuerza = "débil" if abs(correlacion) < 0.3 else ("moderada" if abs(correlacion) < 0.7 else "fuerte")
direccion = "positiva" if correlacion > 0 else "negativa"

resultado = pd.DataFrame({
    'Métrica': ['r de Pearson', 'R²', 'Fuerza', 'Dirección', 'Interpretación'],
    'Valor': [
        f'{correlacion:.3f}',
        f'{r_cuadrado:.3f}',
        fuerza.capitalize(),
        direccion.capitalize(),
        f'GrLivArea explica el {r_cuadrado*100:.1f} % de la varianza de SalePrice'
    ]
})
print(resultado.to_markdown(index=False))
print("\nTable: Correlación de Pearson entre GrLivArea y SalePrice")
```

| Métrica        | Valor                                                   |
|:---------------|:--------------------------------------------------------|
| r de Pearson   | 0.735                                                   |
| R²             | 0.540                                                   |
| Fuerza         | Fuerte                                                  |
| Dirección      | Positiva                                                |
| Interpretación | GrLivArea explica el 54.0 % de la varianza de SalePrice |

Correlación de Pearson entre GrLivArea y SalePrice

Antes de continuar con más análisis, hay una advertencia que todo
analista debe tener siempre presente.

**La correlación NO es causalidad**

Un ejemplo clásico: las ventas de helados y los casos de ahogamiento
tienen una correlación de r = 0.85. ¿Significa que el helado causa
ahogamientos? **¡NO!**

La explicación es que ambas variables aumentan en verano. El calor es
una **variable de confusión** (*confounding variable*) que afecta a
ambas: más calor → más helados + más gente en piscinas.

**SIEMPRE pregúntate antes de asumir causalidad:**

1.  ¿Hay una explicación causal plausible?
2.  ¿Puede haber una tercera variable que explique ambas?
3.  ¿La correlación se mantiene en diferentes subgrupos?

### Limitaciones de la correlación de Pearson

La correlación de Pearson solo detecta relaciones **lineales**. Si la
relación entre dos variables es curva o no lineal, Pearson puede
devolver un valor cercano a cero aunque la asociación sea muy fuerte.

``` python
np.random.seed(42)
x = np.linspace(-10, 10, 100)
y = x**2 + np.random.normal(0, 10, 100)  # parábola con ruido

r_pearson = np.corrcoef(x, y)[0, 1]

demo = pd.DataFrame({
    'Relación': ['Cuadrática (parábola con ruido)'],
    'r de Pearson': [f'{r_pearson:.3f}'],
    'Conclusión incorrecta': ['Sin relación (r ≈ 0)'],
    'Conclusión correcta': ['Relación fuerte, pero no lineal']
})
print(demo.to_markdown(index=False))
print("\nTable: Pearson detecta relación lineal nula aunque la relación cuadrática es evidente")
```

| Relación                        | r de Pearson | Conclusión incorrecta | Conclusión correcta             |
|:---------------------|-----------:|:----------------|:---------------------|
| Cuadrática (parábola con ruido) |        0.012 | Sin relación (r ≈ 0)  | Relación fuerte, pero no lineal |

Pearson detecta relación lineal nula aunque la relación cuadrática es
evidente

### Correlación de Spearman: la alternativa robusta

``` python
from scipy.stats import spearmanr

pearson_r = df['GrLivArea'].corr(df['SalePrice'])
spearman_r, _ = spearmanr(df['GrLivArea'], df['SalePrice'])
diferencia = abs(pearson_r - spearman_r)

comp = pd.DataFrame({
    'Método': ['Pearson', 'Spearman'],
    'Coeficiente (GrLivArea ~ SalePrice)': [round(pearson_r, 3), round(spearman_r, 3)],
})
print(comp.to_markdown(index=False))
conclusion = "Relación aproximadamente lineal — Pearson suficiente" if diferencia <= 0.1 else "Diferencia significativa — investigar con scatterplot"
print(f"\nDiferencia: {diferencia:.3f} → {conclusion}")
print("\nTable: Comparación Pearson vs Spearman — GrLivArea / SalePrice")
```

| Método   | Coeficiente (GrLivArea ~ SalePrice) |
|:---------|------------------------------------:|
| Pearson  |                               0.735 |
| Spearman |                               0.732 |

Diferencia: 0.003 → Relación aproximadamente lineal — Pearson suficiente

Table: Comparación Pearson vs Spearman — GrLivArea / SalePrice

**Análisis de los resultados:** Pearson da 0.735 y Spearman 0.732 para
el par `GrLivArea`-`SalePrice`, con una diferencia de solo 0.003. Al
estar por debajo del umbral de 0,1, la relación es aproximadamente
lineal y Pearson es suficiente en este caso. Si la diferencia hubiera
superado 0,1, habría indicado *outliers* o curvatura que exigirían
investigar con un diagrama de dispersión antes de modelar.

Para un par hipotético, Pearson = 0,40 y Spearman = 0,75 (diferencia =
0,35). ¿Qué conclusión extraes sobre la naturaleza de esa relación? ¿Qué
harías a continuación antes de incluir esa variable en el modelo?

**Cuándo usar cada una:**

**Correlaciones y valores nulos:** `.corr()` omite los nulos en silencio
(*pairwise deletion*): calcula cada par de variables usando solo las
filas donde ambas tienen valor. Si el dataset tiene muchos nulos, dos
correlaciones calculadas sobre muestras muy distintas son difícilmente
comparables. En esta UT los datos ya vienen limpios de UT5, así que no
hay problema — pero tenlo en cuenta cuando trabajes con datos crudos.

Con las correlaciones individuales comprendidas, el siguiente paso es
calcularlas todas a la vez para tener una visión global del dataset.

### Matrices de correlación

``` python
variables_interes = [
    'SalePrice', 'GrLivArea', 'OverallQual',
    'YearBuilt', 'TotalBsmtSF', 'GarageCars'
]
matriz_corr = df[variables_interes].corr(numeric_only=True)

print(matriz_corr.round(2).to_markdown())
print("\nTable: Matriz de correlación — variables numéricas principales (Ames Housing)\n")

corr_con_precio = matriz_corr['SalePrice'].sort_values(ascending=False)
top3 = corr_con_precio[1:4]
top3_df = (
    top3.rename("r").to_frame()
    .assign(varianza_explicada=lambda d: (d["r"] ** 2 * 100).round(1))
)
top3_df.index.name = 'Variable'
print(top3_df.round(3).to_markdown())
print("\nTable: Top 3 predictores de SalePrice — correlación y varianza explicada\n")

umbral = 0.8
corr_vals = (
    matriz_corr
    .where(np.triu(np.ones(matriz_corr.shape, dtype=bool), k=1))
    .stack()
)
pares_multicolineales = corr_vals[corr_vals.abs() > umbral]
if len(pares_multicolineales) == 0:
    print("No se detectan pares de predictores con |r| > 0,80 — sin multicolinealidad grave.")
else:
    mc_df = pares_multicolineales.rename("r").to_frame().reset_index()
    mc_df.columns = ['Var 1', 'Var 2', 'r']
    print(mc_df.round(3).to_markdown(index=False))
    print("\nTable: Pares con multicolinealidad (|r| > 0,80)")
```

|             | SalePrice | GrLivArea | OverallQual | YearBuilt | TotalBsmtSF | GarageCars |
|:---------|---------:|---------:|-----------:|---------:|-----------:|----------:|
| SalePrice   |         1 |      0.73 |         0.8 |      0.52 |        0.65 |       0.64 |
| GrLivArea   |      0.73 |         1 |        0.59 |      0.19 |        0.41 |       0.48 |
| OverallQual |       0.8 |      0.59 |           1 |      0.57 |        0.54 |        0.6 |
| YearBuilt   |      0.52 |      0.19 |        0.57 |         1 |         0.4 |       0.54 |
| TotalBsmtSF |      0.65 |      0.41 |        0.54 |       0.4 |           1 |       0.45 |
| GarageCars  |      0.64 |      0.48 |         0.6 |      0.54 |        0.45 |          1 |

Matriz de correlación — variables numéricas principales (Ames Housing)

| Variable    |     r | varianza_explicada |
|:------------|------:|-------------------:|
| OverallQual | 0.796 |               63.3 |
| GrLivArea   | 0.735 |                 54 |
| TotalBsmtSF | 0.651 |               42.4 |

Top 3 predictores de SalePrice — correlación y varianza explicada

No se detectan pares de predictores con \|r\| \> 0,80 — sin
multicolinealidad grave.

**Análisis de los resultados:** Los principales predictores de
`SalePrice` son `OverallQual` (r = 0.796), `GrLivArea` (r = 0.735),
`TotalBsmtSF` (r = 0.651), `GarageCars` (r = 0.641), `YearBuilt` (r =
0.524). Estas variables explican conjuntamente gran parte de la varianza
del precio. No se detectan pares con \|r\| \> 0.8 — sin
multicolinealidad grave entre estos predictores. Pueden incluirse todas
en un modelo sin riesgo de coeficientes inestables.

**Manos a la Obra**

#### Análisis de correlaciones y multicolinealidad

**Objetivo:** identificar las variables más predictivas y detectar
redundancias.

**Contexto profesional:** Necesitas seleccionar las mejores variables
para un modelo de predicción de precios. Tu jefe te pidió reducir las 74
variables disponibles a 10 sin perder poder predictivo.

**Instrucciones (Autónomo):**

Trabaja con estas 12 variables numéricas: `SalePrice`, `GrLivArea`,
`TotalBsmtSF`, `1stFlrSF`, `GarageArea`, `GarageCars`, `TotRmsAbvGrd`,
`OverallQual`, `YearBuilt`, `YearRemodAdd`, `MasVnrArea`, `LotArea`.

1.  Calcula la matriz de correlación completa (Pearson).
2.  Identifica las 5 variables más correlacionadas con `SalePrice`.
3.  Detecta pares con multicolinealidad (`|r| > 0.75`). Para cada par
    problemático, decide cuál eliminar y justifica.
4.  Compara Pearson vs. Spearman en las 3 variables con mayor
    correlación con `SalePrice`.

`GarageCars` y `GarageArea` tienen $r = 0{,}88$ entre sí. ¿Cuál
eliminarías del modelo y por qué? ¿Cambia tu respuesta si sabes que
`GarageCars` es más fácil de recopilar en casas nuevas?

Comparaste Pearson vs. Spearman para `OverallQual`: Spearman = 0,80 y
Pearson = 0,75. ¿Qué te indica esa diferencia sobre la naturaleza de la
relación entre `OverallQual` y `SalePrice`? ¿Cuál de los dos
coeficientes es más apropiado para esta variable?

**Criterio de éxito:**

-   Matriz de correlación generada y legible.
-   Has identificado correctamente multicolinealidad.
-   Puedes justificar qué variable conservar de cada par.
-   Has comparado Pearson vs. Spearman en al menos 3 casos.

La matriz de correlación es tu mejor aliada para detectar redundancias.
Tienes los comandos necesarios en el **Apéndice C.2**.

``` python
# Escribe aquí tu código
```

Las correlaciones numéricas cubren relaciones entre variables continuas.
Para entender cómo influyen las variables categóricas — barrio, calidad,
tipo de garaje — en el precio, necesitamos un enfoque diferente.

### Categórica vs. numérica: análisis por grupos con `groupby`

``` python
precio_por_barrio = df.groupby('Neighborhood')['SalePrice'].agg([
    ('n', 'count'),
    ('Promedio', 'mean'),
    ('Mediana', 'median'),
    ('Desv. Std', 'std'),
]).sort_values('Promedio', ascending=False)

precio_por_barrio[['Promedio', 'Mediana', 'Desv. Std']] = \
    precio_por_barrio[['Promedio', 'Mediana', 'Desv. Std']].round(0).astype(int)

print(precio_por_barrio.head(10).to_markdown())
print("\nTable: Top 10 barrios por precio medio — Neighborhood en Ames Housing\n")

promedio_global = df['SalePrice'].mean()
precio_por_barrio['vs_media (%)'] = \
    ((precio_por_barrio['Promedio'] - promedio_global) / promedio_global * 100).round(1)

premium = precio_por_barrio[precio_por_barrio['vs_media (%)'] > 25][['Promedio', 'vs_media (%)']]
economicos = precio_por_barrio[precio_por_barrio['vs_media (%)'] < -25][['Promedio', 'vs_media (%)']]

print(premium.to_markdown())
print("\nTable: Barrios premium — precio >25% sobre la media global\n")

print(economicos.to_markdown())
print("\nTable: Barrios económicos — precio <25% por debajo de la media global")
```

| Neighborhood |   n | Promedio | Mediana | Desv. Std |
|:-------------|----:|---------:|--------:|----------:|
| NoRidge      |  41 |   335295 |  301500 |    121413 |
| NridgHt      |  77 |   316271 |  315000 |     96393 |
| StoneBr      |  25 |   310499 |  278000 |    112970 |
| Timber       |  38 |   242247 |  228475 |     64846 |
| Veenker      |  11 |   238773 |  218000 |     72369 |
| Somerst      |  86 |   225380 |  225500 |     56178 |
| ClearCr      |  28 |   212565 |  200250 |     50232 |
| Crawfor      |  51 |   210625 |  200624 |     68866 |
| CollgCr      | 150 |   197966 |  197200 |     51404 |
| Blmngtn      |  17 |   194871 |  191000 |     30393 |

Top 10 barrios por precio medio — Neighborhood en Ames Housing

| Neighborhood | Promedio | vs_media (%) |
|:-------------|---------:|-------------:|
| NoRidge      |   335295 |         85.3 |
| NridgHt      |   316271 |         74.8 |
| StoneBr      |   310499 |         71.6 |
| Timber       |   242247 |         33.9 |
| Veenker      |   238773 |           32 |

Barrios premium — precio \>25% sobre la media global

| Neighborhood | Promedio | vs_media (%) |
|:-------------|---------:|-------------:|
| OldTown      |   128225 |        -29.1 |
| Edwards      |   127319 |        -29.6 |
| BrkSide      |   124834 |          -31 |
| BrDale       |   104494 |        -42.2 |
| IDOTRR       |   100124 |        -44.7 |
| MeadowV      |    98576 |        -45.5 |

Barrios económicos — precio \<25% por debajo de la media global

**Análisis de los resultados:** El precio por barrio varía de forma
dramática: NoRidge (\$335.295), NridgHt (\$316.271) y StoneBr
(\$310.499) lideran el segmento *premium*, mientras que MeadowV
(\$98.576) queda en el extremo opuesto. La ratio máximo/mínimo es de
3,4x, lo que confirma que `Neighborhood` es una de las variables
categóricas con mayor poder predictivo del *dataset*. Un modelo que no
incluya el barrio como variable estará sistemáticamente sesgado:
infraestimará precios en zonas *premium* y los sobreestimará en las
económicas.

No te quedes solo con la media. La desviación estándar te dice si los
precios en ese barrio son homogéneos o muy variables. Alta *std* =
barrio mixto (casas de diferentes calidades).

El análisis por grupos revela diferencias entre barrios, pero no nos
dice si dentro de cada barrio la calidad de construcción sigue el mismo
patrón. Para cruzar dos variables categóricas necesitamos una tabla de
contingencia.

### Tablas de contingencia simples

``` python
tabla_contingencia = pd.crosstab(
    df['Neighborhood'],
    df['OverallQual'],
    margins=True,
    margins_name='Total'
)
print(tabla_contingencia.head(10).to_markdown())
print("\nTable: Distribución de OverallQual por barrio — top 10 (con totales)\n")

tabla_proporciones = pd.crosstab(
    df['Neighborhood'],
    df['OverallQual'],
    normalize='index'
)
print((tabla_proporciones * 100).round(1).head().to_markdown())
print("\nTable: Proporción (%) de cada nivel de calidad dentro del barrio — top 5\n")

barrios_calidad_alta = tabla_proporciones[
    (tabla_proporciones.get(8, 0) + tabla_proporciones.get(9, 0) + tabla_proporciones.get(10, 0)) > 0.5
]
print(f"Barrios donde >50% tienen calidad 8-10: **{', '.join(barrios_calidad_alta.index.tolist())}**")
```

| Neighborhood |   1 |   2 |   3 |   4 |   5 |   6 |   7 |   8 |   9 |  10 | Total |
|:-------------|----:|----:|----:|----:|----:|----:|----:|----:|----:|----:|------:|
| Blmngtn      |   0 |   0 |   0 |   0 |   0 |   0 |  14 |   3 |   0 |   0 |    17 |
| Blueste      |   0 |   0 |   0 |   0 |   0 |   2 |   0 |   0 |   0 |   0 |     2 |
| BrDale       |   0 |   0 |   0 |   0 |   5 |  11 |   0 |   0 |   0 |   0 |    16 |
| BrkSide      |   1 |   1 |   3 |  10 |  21 |  18 |   4 |   0 |   0 |   0 |    58 |
| ClearCr      |   0 |   0 |   0 |   3 |   6 |  10 |   9 |   0 |   0 |   0 |    28 |
| CollgCr      |   0 |   0 |   0 |   1 |  28 |  24 |  71 |  23 |   3 |   0 |   150 |
| Crawfor      |   0 |   0 |   0 |   1 |  11 |  19 |  13 |   7 |   0 |   0 |    51 |
| Edwards      |   1 |   0 |   5 |  25 |  38 |  23 |   5 |   0 |   1 |   0 |    98 |
| Gilbert      |   0 |   0 |   0 |   1 |   3 |  33 |  36 |   5 |   1 |   0 |    79 |
| IDOTRR       |   0 |   1 |   3 |  12 |  11 |   8 |   2 |   0 |   0 |   0 |    37 |

Distribución de OverallQual por barrio — top 10 (con totales)

| Neighborhood |   1 |   2 |   3 |    4 |    5 |    6 |    7 |    8 |   9 |  10 |
|:-------------|----:|----:|----:|-----:|-----:|-----:|-----:|-----:|----:|----:|
| Blmngtn      |   0 |   0 |   0 |    0 |    0 |    0 | 82.4 | 17.6 |   0 |   0 |
| Blueste      |   0 |   0 |   0 |    0 |    0 |  100 |    0 |    0 |   0 |   0 |
| BrDale       |   0 |   0 |   0 |    0 | 31.2 | 68.8 |    0 |    0 |   0 |   0 |
| BrkSide      | 1.7 | 1.7 | 5.2 | 17.2 | 36.2 |   31 |  6.9 |    0 |   0 |   0 |
| ClearCr      |   0 |   0 |   0 | 10.7 | 21.4 | 35.7 | 32.1 |    0 |   0 |   0 |

Proporción (%) de cada nivel de calidad dentro del barrio — top 5

Barrios donde \>50% tienen calidad 8-10: **NoRidge, NridgHt, StoneBr**

**Análisis de los resultados:** La tabla cruzada confirma que la calidad
no se distribuye de forma uniforme entre barrios: `NoRidge`, `NridgHt`,
`StoneBr` concentran sus viviendas en los niveles 8-10, mientras que
`MeadowV`, `Sawyer` tienen la mayoría en los niveles 4 y 5. Exactamente
3 barrios superan el umbral del 50 % de viviendas en calidad 8-10, y
coinciden con los que lideran el precio medio. Esto confirma que
`Neighborhood` y `OverallQual` no son variables independientes: hay
barrios donde la calidad alta es la norma, y otros donde es la
excepción.

**Manos a la Obra**

#### Análisis de segmentación por categorías

**Objetivo:** descubrir cómo variables categóricas impactan al *target*
y detectar segmentos de mercado.

**Contexto profesional:** Eres analista de una inmobiliaria y necesitas
identificar segmentos de mercado para una campaña de marketing dirigida.
El CMO quiere saber: “¿qué características definen el mercado *premium*
vs. económico?”.

**Instrucciones (Autónomo):**

Realiza análisis de grupos para `Neighborhood` (barrio), `HouseStyle`
(estilo de casa) y `OverallQual` (calidad general):

1.  Para `Neighborhood`: calcula *n*, media, mediana, *std*, min y max.
    Identifica el *top* 5 de barrios *premium* y *bottom* 5 económicos.
    Calcula la desviación de cada barrio respecto al promedio global.
2.  Para `HouseStyle`: identifica los estilos con mayor variabilidad de
    precio (mayor CV).
3.  Para `OverallQual`: analiza si la relación con `SalePrice` es lineal
    e identifica el umbral de calidad donde el precio se dispara.
4.  Crea una tabla de contingencia `Neighborhood` vs. `OverallQual`.
    Identifica barrios donde predominan casas de baja calidad.

“OldTown” tiene un precio medio de 120.000 € con $s = 40.000$ €, y
“StoneBr” tiene 300.000 € con $s = 80.000$ €. ¿Cuál de los dos barrios
es más homogéneo en precios? Calcula el CV de cada uno para justificar
tu respuesta.

Los estilos “1Story” y “2Story” tienen precios medios similares pero
medianas muy diferentes. ¿Qué te dice esa discrepancia? ¿Cuál de las dos
medidas (media o mediana) usarías para comunicar el “precio típico” de
cada estilo al CMO?

**Criterio de éxito:**

-   Has usado `.groupby().agg()` con múltiples funciones.
-   Has calculado la desviación porcentual vs. promedio global.
-   Has identificado segmentos de mercado claramente diferenciados.
-   Has generado al menos una tabla de contingencia interpretable.

El análisis por grupos con `.groupby()` es fundamental para segmentar el
mercado. Tienes las funciones de agregación en el **Apéndice C.2**.

``` python
# Escribe aquí tu código
```

In [42]:
# ESTRUCTURA SUGERIDA (sin código completo):
resultados = df.groupby('Variable_Categorica')['SalePrice'].agg({
    'n': 'count',
    'promedio': 'mean',
    # ... añade más según necesites
})

Al presentar resultados a *stakeholders*, ordena siempre la tabla por
promedio — descendente para destacar los mejores, ascendente para los
peores. Una tabla ordenada alfabéticamente no comunica el *insight* de
qué barrios son *premium*.

Hasta ahora hemos cruzado una categórica con una numérica. Cuando ambas
variables son categóricas, el crosstab es la herramienta adecuada.

### Categórica vs. categórica: tablas de contingencia

``` python
tabla = pd.crosstab(df['GarageType'], df['HouseStyle'], margins=True)
print(tabla.to_markdown())
print("\nTable: Tabla de contingencia GarageType × HouseStyle\n")

tabla_prop = pd.crosstab(df['GarageType'], df['HouseStyle'], normalize='index')
print((tabla_prop * 100).round(1).to_markdown())
print("\nTable: Proporción (%) de estilos arquitectónicos por tipo de garaje")
```

| GarageType | 1.5Fin | 1.5Unf | 1Story | 2.5Fin | 2.5Unf | 2Story | SFoyer | SLvl |  All |
|:-----------|-------:|-------:|-------:|-------:|-------:|-------:|-------:|-----:|-----:|
| 2Types     |      0 |      0 |      4 |      0 |      0 |      2 |      0 |    0 |    6 |
| Attchd     |     57 |      5 |    550 |      3 |      1 |    274 |     20 |   40 |  950 |
| Basment    |      1 |      0 |      6 |      1 |      0 |      2 |      2 |    7 |   19 |
| BuiltIn    |      5 |      0 |      1 |      0 |      0 |     68 |      1 |   12 |   87 |
| CarPort    |      1 |      0 |      5 |      0 |      0 |      2 |      1 |    0 |    9 |
| Detchd     |     90 |      9 |    160 |      4 |     10 |     95 |     13 |    6 |  387 |
| All        |    154 |     14 |    726 |      8 |     11 |    443 |     37 |   65 | 1458 |

Tabla de contingencia GarageType × HouseStyle

| GarageType | 1.5Fin | 1.5Unf | 1Story | 2.5Fin | 2.5Unf | 2Story | SFoyer | SLvl |
|:-----------|-------:|-------:|-------:|-------:|-------:|-------:|-------:|-----:|
| 2Types     |      0 |      0 |   66.7 |      0 |      0 |   33.3 |      0 |    0 |
| Attchd     |      6 |    0.5 |   57.9 |    0.3 |    0.1 |   28.8 |    2.1 |  4.2 |
| Basment    |    5.3 |      0 |   31.6 |    5.3 |      0 |   10.5 |   10.5 | 36.8 |
| BuiltIn    |    5.7 |      0 |    1.1 |      0 |      0 |   78.2 |    1.1 | 13.8 |
| CarPort    |   11.1 |      0 |   55.6 |      0 |      0 |   22.2 |   11.1 |    0 |
| Detchd     |   23.3 |    2.3 |   41.3 |      1 |    2.6 |   24.5 |    3.4 |  1.6 |

Proporción (%) de estilos arquitectónicos por tipo de garaje

**Análisis de los resultados:** Las proporciones varían notablemente
entre filas: los garajes de tipo `Attchd` (integrado en la vivienda)
aparecen en proporciones mucho más altas en casas de 2 plantas
(`2Story`) que en las de 1 planta (`1Story`), mientras que los garajes
`Detchd` (independientes) son relativamente más frecuentes en los
estilos más antiguos. Esa variación entre filas es la señal de
**asociación**: el tipo de garaje y el estilo arquitectónico no son
independientes, probablemente porque ambos reflejan la época de
construcción. En la práctica esto tiene implicaciones para el modelado:
incluir ambas variables puede introducir redundancia parcial.

### Concepto intuitivo de independencia

``` python
tabla_prop = pd.crosstab(df['CentralAir'], df['Heating'], normalize='index')
print((tabla_prop * 100).round(2).to_markdown())
print("\nTable: Proporción (%) de tipos de calefacción por presencia de aire acondicionado central\n")

variabilidad = tabla_prop.std().rename("Std proporciones")
etiquetas = variabilidad.apply(
    lambda v: "posible independencia" if v < 0.05 else "hay asociación"
).rename("Interpretación")
print(pd.concat([variabilidad.round(4), etiquetas], axis=1).to_markdown())
print("\nTable: Variabilidad entre grupos — indicador de asociación o independencia")
```

| CentralAir | Floor |  GasA |  GasW | Grav | OthW | Wall |
|:-----------|------:|------:|------:|-----:|-----:|-----:|
| N          |  1.05 | 72.63 | 12.63 | 7.37 | 2.11 | 4.21 |
| Y          |     0 | 99.56 |  0.44 |    0 |    0 |    0 |

Proporción (%) de tipos de calefacción por presencia de aire
acondicionado central

| Heating | Std proporciones | Interpretación        |
|:--------|-----------------:|:----------------------|
| Floor   |           0.0074 | posible independencia |
| GasA    |           0.1904 | hay asociación        |
| GasW    |           0.0862 | hay asociación        |
| Grav    |           0.0521 | hay asociación        |
| OthW    |           0.0149 | posible independencia |
| Wall    |           0.0298 | posible independencia |

Variabilidad entre grupos — indicador de asociación o independencia

**Análisis de los resultados:** Las proporciones de tipo de calefacción
difieren notablemente entre viviendas con y sin aire acondicionado
central: las viviendas con `CentralAir=Y` concentran una proporción muy
superior en calefacción de gas (`GasA`), mientras que en las de
`CentralAir=N` aparecen con más frecuencia sistemas alternativos. La
desviación estándar de las proporciones por columna confirma esta
variación, señalando **asociación** entre ambas variables. En la
práctica, incluir las dos en un modelo puede introducir redundancia
parcial: ambas reflejan, en parte, la modernidad de las instalaciones de
la vivienda.

El análisis de contingencia muestra que `CentralAir` y `Heating` están
asociados. Si incluyes ambas variables en un modelo de regresión, ¿qué
problema técnico concreto puedes provocar? ¿Cómo decidirías cuál de las
dos conservar?

**Manos a la Obra**

#### Análisis de independencia entre categóricas

**Objetivo:** Determinar mediante análisis de contingencia si existe una
asociación estadística relevante entre diferentes características
cualitativas de las viviendas.

**Contexto profesional:** Como parte de un proceso de selección de
características (*feature selection*), debes identificar variables
categóricas redundantes (altamente asociadas) para simplificar el modelo
final y evitar multicolinealidad cualitativa.

**Instrucciones (Autónomo):**

1.  Selecciona dos pares de variables categóricas que creas que podrían
    estar relacionadas (ej: `GarageType` vs `GarageFinish`) y otro par
    que creas que debería ser independiente.
2.  Genera las tablas de contingencia normalizadas por filas para ambos
    casos.
3.  **Juicio profesional:** Analiza los resultados y determina,
    basándote exclusivamente en los datos obtenidos, qué par presenta
    una asociación real y cuál parece ser independiente. Demuestra tu
    conclusión señalando cómo varían las proporciones en las celdas de
    la tabla.
4.  Justifica cuál de las variables eliminarías del modelo si
    encontraras una asociación superior al 90%.

**Criterio de éxito:** Presentar dos tablas de contingencia legibles y
una conclusión escrita que identifique correctamente la presencia o
ausencia de asociación basada en la variabilidad de las frecuencias
relativas. Referencia: si usas `GarageType` vs `GarageFinish`, el tipo
`Detchd` (garaje independiente) muestra un 91.5 % sin acabar (`Unf`),
mientras que `BuiltIn` (integrado) muestra un 66.7 % acabado (`Fin`) —
variabilidad entre filas alta (std \> 0.30), asociación clara. Para el
par independiente, busca uno donde la std de proporciones por fila sea
\< 0.05.

In [45]:
# Escribe tu codigo aqui


Para 2-3 pares de variables categóricas, el análisis visual de
proporciones es suficiente y más interpretable que cualquier test
estadístico. Cuando el número de combinaciones crece, prioriza los pares
que tienen sentido de negocio — no analices todas las combinaciones
posibles.

### Consolidación: análisis bivariante

## Análisis multivariante: explorando interacciones

La forma de detectar estas interacciones es estratificar: analizar la
relación por separado dentro de cada subgrupo.

### Estratificación: análisis condicional

``` python
barrios_validos = df['Neighborhood'].value_counts()
barrios_validos = barrios_validos[barrios_validos >= 20].index

def corr_grlivarea_saleprice(sub):
    return sub['GrLivArea'].corr(sub['SalePrice'])

df_resultados = (
    df[df['Neighborhood'].isin(barrios_validos)]
    .groupby('Neighborhood')
    .apply(corr_grlivarea_saleprice, include_groups=False)
    .rename("Correlación")
    .to_frame()
    .assign(n=df[df['Neighborhood'].isin(barrios_validos)].groupby('Neighborhood').size())
    .sort_values('Correlación', ascending=False)
)
df_resultados['Correlación'] = df_resultados['Correlación'].round(3)
print(df_resultados.to_markdown())
print("\nTable: Correlación GrLivArea–SalePrice estratificada por barrio (n ≥ 20)\n")

rango_corr = df_resultados['Correlación'].max() - df_resultados['Correlación'].min()
hay_interaccion = rango_corr > 0.30
conclusion = "HAY INTERACCIÓN — la relación varía según el barrio" if hay_interaccion else "Sin interacción — relación consistente"
print(f"Rango: {df_resultados['Correlación'].min():.3f} – {df_resultados['Correlación'].max():.3f} "
      f"(amplitud {rango_corr:.3f}) → **{conclusion}**")
```

| Neighborhood | Correlación |   n |
|:-------------|------------:|----:|
| NoRidge      |       0.887 |  41 |
| CollgCr      |       0.865 | 150 |
| BrkSide      |       0.835 |  58 |
| SawyerW      |       0.819 |  59 |
| StoneBr      |       0.811 |  25 |
| SWISU        |       0.793 |  25 |
| NWAmes       |       0.765 |  73 |
| NridgHt      |        0.75 |  77 |
| OldTown      |       0.695 | 113 |
| Somerst      |       0.692 |  86 |
| NAmes        |       0.679 | 225 |
| Crawfor      |       0.676 |  51 |
| Edwards      |       0.626 |  98 |
| Timber       |       0.601 |  38 |
| Gilbert      |       0.581 |  79 |
| Mitchel      |       0.577 |  49 |
| IDOTRR       |       0.561 |  37 |
| Sawyer       |        0.53 |  74 |
| ClearCr      |        0.32 |  28 |

Correlación GrLivArea–SalePrice estratificada por barrio (n ≥ 20)

Rango: 0.320 – 0.887 (amplitud 0.567) → **HAY INTERACCIÓN — la relación
varía según el barrio**

**Análisis de los resultados:** La correlación entre `GrLivArea` y
`SalePrice` varía de 0.320 en `ClearCr` a 0.887 en `NoRidge` — una
amplitud de 0.567. En zonas de lujo, cada metro cuadrado adicional se
traduce en un incremento de precio significativamente mayor que en zonas
populares. Esta interacción es una señal de *feature engineering*:
incluir un término de interacción `GrLivArea × Neighborhood` (o trabajar
con un modelo de árboles que la captura automáticamente) mejorará la
precisión de la predicción.

La correlación entre `GrLivArea` y `SalePrice` varía entre 0.320 en
`ClearCr` y 0.887 en `NoRidge`. ¿Por qué un modelo lineal sin esta
interacción subestimaría el precio en barrios *premium* y lo
sobreestimaría en los económicos? ¿Qué tipo de modelo capturaría esta
interacción de forma automática sin necesidad de crearla manualmente?

El análisis de interacciones nos lleva al siguiente nivel: en lugar de
estudiar barrio y calidad por separado, podemos cruzarlos
simultáneamente con una *pivot table*.

### Agregaciones múltiples y *pivoting*

``` python
pivot = df.pivot_table(
    values='SalePrice',
    index='Neighborhood',
    columns='OverallQual',
    aggfunc='mean'
)
barrios_relevantes = df['Neighborhood'].value_counts().head(8).index
calidades_relevantes = [5, 6, 7, 8, 9, 10]
pivot_filtrado = pivot.loc[barrios_relevantes, calidades_relevantes].round(0)
print(pivot_filtrado.to_markdown())
print("\nTable: Precio medio (USD) por barrio × calidad — top 8 barrios, calidades 5-10\n")

def analizar_patron(serie):
    incrementos = serie.dropna().sort_index().diff().dropna()
    if len(incrementos) < 2:
        return pd.Series({'inc_promedio': float('nan'), 'std_inc': float('nan'), 'no_lineal': False})
    media = incrementos.mean()
    std = incrementos.std()
    return pd.Series({'inc_promedio': round(media, 0), 'std_inc': round(std, 0),
                      'no_lineal': bool((std / media > 0.5) if media else False)})

patrones = pivot.loc[barrios_relevantes].apply(analizar_patron, axis=1)
print(patrones.to_markdown())
print("\nTable: Patrón de incremento de precio por punto de calidad — por barrio")
```

| Neighborhood |      5 |      6 |      7 |      8 |      9 |     10 |
|:-------------|-------:|-------:|-------:|-------:|-------:|-------:|
| NAmes        | 136949 | 155057 | 203933 | 282667 |    nan |    nan |
| CollgCr      | 134859 | 170101 | 209840 | 259876 | 279000 |    nan |
| OldTown      | 116244 | 129837 | 166580 | 150333 |    nan | 400000 |
| Edwards      | 123568 | 149325 | 152280 |    nan | 320000 |    nan |
| Somerst      |    nan | 184048 | 203685 | 255626 | 320978 | 310000 |
| Gilbert      | 168713 | 181464 | 200724 | 199300 | 377500 |    nan |
| NridgHt      |    nan | 195350 | 210308 | 299461 | 379259 | 426195 |
| Sawyer       | 136432 | 145582 |    nan |    nan |    nan |    nan |

Precio medio (USD) por barrio × calidad — top 8 barrios, calidades 5-10

| Neighborhood | inc_promedio | std_inc | no_lineal |
|:-------------|-------------:|--------:|:----------|
| NAmes        |        38273 |   25664 | True      |
| CollgCr      |        31080 |   15712 | True      |
| OldTown      |        50237 |   99330 | True      |
| Edwards      |        43167 |   61701 | True      |
| Somerst      |        31488 |   34200 | True      |
| Gilbert      |        47300 |   73947 | True      |
| NridgHt      |        57711 |   33765 | True      |
| Sawyer       |         8682 |     661 | False     |

Patrón de incremento de precio por punto de calidad — por barrio

**Análisis de los resultados:** La columna `incremento_promedio` indica
cuántos euros añade de media cada punto extra de calidad en ese barrio.
En NridgHt cada punto vale ~57.700 €, mientras que en Sawyer solo ~8.700
€: la calidad “pesa” 6,6 veces más en el barrio *premium*. La columna
`std_incremento` mide si ese incremento es regular o errático: en
Edwards la desviación (91.503 €) supera al incremento medio (15.911 €),
lo que indica saltos muy irregulares — posiblemente por pocas
observaciones en alguna categoría o por NaN en la *pivot table*. La
columna `no_lineal` resume ambas señales: todos los barrios con muestra
suficiente muestran no linealidad. Eso confirma que un modelo lineal que
trate calidad y barrio como efectos independientes estará mal
especificado; necesitarás una variable de interacción
`OverallQual × Neighborhood` o un modelo basado en árboles que la
capture automáticamente.

La *pivot table* muestra que en NridgHt cada punto de `OverallQual`
añade ~57.700 € al precio, mientras que en Sawyer solo añade ~8.700 €.
¿Qué *feature* de interacción crearías para capturar este patrón en un
modelo de regresión lineal? ¿Por qué un árbol de decisión no necesitaría
esa *feature* adicional?

### *Pivot tables* avanzadas: múltiples agregaciones

``` python
pivot_completo = df.pivot_table(
    values='SalePrice',
    index='Neighborhood',
    columns='OverallQual',
    aggfunc=['mean', 'count', 'std']
)

print(pivot_completo['mean'].round(0).head().to_markdown())
print("\nTable: Precio medio (USD) por barrio × calidad — primeros 5 barrios\n")

print(pivot_completo['count'].head().to_markdown())
print("\nTable: Número de observaciones por celda (barrio × calidad) — primeros 5 barrios\n")

conteos = pivot_completo['count']
celdas_problematicas = conteos < 5
n_prob = int(celdas_problematicas.sum().sum())
pct_prob = n_prob / conteos.size * 100
print(f"Combinaciones totales: {conteos.size} | Con n<5: **{n_prob}** ({pct_prob:.1f}%) — revisar antes de modelar")
```

| Neighborhood |     1 |     2 |     3 |      4 |      5 |      6 |      7 |      8 |   9 |  10 |
|:-----------|-----:|-----:|-----:|------:|------:|------:|------:|------:|----:|-----:|
| Blmngtn      |   nan |   nan |   nan |    nan |    nan |    nan | 188660 | 223854 | nan | nan |
| Blueste      |   nan |   nan |   nan |    nan |    nan | 137500 |    nan |    nan | nan | nan |
| BrDale       |   nan |   nan |   nan |    nan | 117000 |  98809 |    nan |    nan | nan | nan |
| BrkSide      | 39300 | 60000 | 67000 | 100850 | 124662 | 145843 | 172125 |    nan | nan | nan |
| ClearCr      |   nan |   nan |   nan | 183977 | 173083 | 209850 | 251433 |    nan | nan | nan |

Precio medio (USD) por barrio × calidad — primeros 5 barrios

| Neighborhood |   1 |   2 |   3 |   4 |   5 |   6 |   7 |   8 |   9 |  10 |
|:-------------|----:|----:|----:|----:|----:|----:|----:|----:|----:|----:|
| Blmngtn      | nan | nan | nan | nan | nan | nan |  14 |   3 | nan | nan |
| Blueste      | nan | nan | nan | nan | nan |   2 | nan | nan | nan | nan |
| BrDale       | nan | nan | nan | nan |   5 |  11 | nan | nan | nan | nan |
| BrkSide      |   1 |   1 |   3 |  10 |  21 |  18 |   4 | nan | nan | nan |
| ClearCr      | nan | nan | nan |   3 |   6 |  10 |   9 | nan | nan | nan |

Número de observaciones por celda (barrio × calidad) — primeros 5
barrios

Combinaciones totales: 250 \| Con n\<5: **43** (17.2%) — revisar antes
de modelar

**Análisis de los resultados:** Con 250 combinaciones posibles (25
barrios × 10 niveles de calidad) y solo 1.458 viviendas, era inevitable
que muchas celdas quedaran vacías o con muy pocos casos. El output lo
confirma: el 17,6 % de las celdas tienen n \< 5 y muchas directamente
son NaN. Eso no invalida el análisis, pero sí obliga a interpretarlo con
cautela: un promedio basado en 2 o 3 observaciones es muy sensible a
outliers. La tabla de conteos es tan importante como la de promedios —
léelas siempre en paralelo. Si el porcentaje problemático superara el
30-40 %, lo adecuado sería reducir la granularidad: por ejemplo, agrupar
los 10 niveles de calidad en tres categorías (baja 1-4, media 5-7, alta
8-10) antes de pivotar.

**Manos a la Obra**

#### Análisis de interacciones multivariante

**Objetivo:** Detectar interacciones entre variables que no son visibles
mediante análisis bivariante simple y cuantificar su impacto en el
precio.

**Contexto profesional:** Tu director comercial te plantea una
hipótesis: *“Creo que los metros cuadrados no se valoran igual en un
barrio humilde que en uno de lujo”*. Tu tarea es demostrar o desmentir
esta hipótesis con evidencias estadísticas.

**Instrucciones (Autónomo):**

1.  Selecciona los **3 barrios con más registros** del dataset.
2.  Calcula la correlación entre la superficie habitable (`GrLivArea`) y
    el precio (`SalePrice`) de forma independiente para cada uno de esos
    3 barrios.
3.  Crea una tabla dinámica (`pivot_table`) que muestre el precio medio
    cruzando el barrio (`Neighborhood`) con la calidad general
    (`OverallQual`) para los barrios seleccionados.
4.  **Informe de hallazgos:** ¿Es constante la relación entre metros
    cuadrados y precio en los tres barrios? Identifica una combinación
    de Barrio/Calidad donde el precio medio sea inusualmente alto o bajo
    en comparación con la tendencia general.

**Criterio de éxito:** Presentar las 3 correlaciones diferenciadas y la
tabla dinámica. La conclusión debe responder explícitamente a la
hipótesis del director comercial usando los datos obtenidos.

In [51]:
# Escribe tu codigo aqui


En modelos tree-based (Random Forest, XGBoost), las interacciones se
capturan automáticamente. En regresión lineal, debes crearlas
manualmente. Esto afecta tu estrategia de *feature engineering*.

### Consolidación: análisis multivariante

## Estadística inferencial: para ir más allá

**Para ir más allá — Tests de hipótesis y p-valores**

El EDA que has aprendido en esta UT cubre el análisis exploratorio de
patrones. Cuando necesites **confirmar estadísticamente** si una
diferencia entre grupos es real o podría ser fruto del azar (A/B
testing, validación de modelos, investigación), entra en juego la
*estadística inferencial*.

Las herramientas clave están en `scipy.stats`: `ttest_ind()` para
comparar medias de dos grupos, `mannwhitneyu()` como alternativa no
paramétrica, y `chi2_contingency()` para variables categóricas. Para el
80% de los proyectos en tu primer empleo, el EDA de esta UT es
suficiente.

## Metodología de EDA: flujo de trabajo profesional

El flujo detallado con las cinco fases, tiempos orientativos y la
plantilla de *notebook* están en el **Apéndice — Guía rápida de
referencia, sección UT6**.

### Registro de hallazgos clave

**Buena práctica:** Mantén un “notebook de hallazgos” paralelo donde
documentes los resultados clave. El siguiente bloque genera
automáticamente ese registro a partir de los datos reales:

**1. Variables más predictivas (\|r\| con SalePrice)**

|             |       |
|:------------|------:|
| OverallQual | 0.796 |
| GrLivArea   | 0.735 |
| TotalBsmtSF | 0.651 |
| GarageCars  | 0.641 |
| 1stFlrSF    | 0.632 |

Top 5 variables con mayor correlación con SalePrice

**2. Candidatas a transformación logarítmica (skew \> 1)**

|             |   skew | skew_log | recomendacion |
|:------------|-------:|---------:|:--------------|
| SalePrice   |  +1.88 |    +0.12 | aplicar log   |
| LotArea     | +12.57 |    -0.18 | aplicar log   |
| GrLivArea   |  +1.01 |    -0.07 | aplicar log   |
| TotalBsmtSF |  +0.51 |    -5.17 | mantener      |
| 1stFlrSF    |  +0.89 |    +0.01 | mantener      |

Asimetría antes y después de log1p — candidatas a transformación

**3. Multicolinealidad detectada (\|r\| \> 0.8 entre predictores)**

| var1        | var2         |       |
|:------------|:-------------|------:|
| GarageCars  | GarageArea   | 0.887 |
| GrLivArea   | TotRmsAbvGrd | 0.829 |
| TotalBsmtSF | 1stFlrSF     | 0.804 |

Pares de predictores con correlación superior a 0.80

**4. Categorías raras a agrupar (n \< 10)**

| columna      | categoría |   n |
|:-------------|:----------|----:|
| Neighborhood | NPkVill   |   9 |
| Neighborhood | Blueste   |   2 |
| HouseStyle   | 2.5Fin    |   8 |
| Condition1   | PosA      |   8 |
| Condition1   | RRNn      |   5 |
| Condition1   | RRNe      |   2 |
| SaleType     | ConLD     |   9 |
| SaleType     | ConLI     |   5 |
| SaleType     | ConLw     |   5 |
| SaleType     | CWD       |   4 |
| SaleType     | Oth       |   3 |
| SaleType     | Con       |   2 |

Categorías con menos de 10 observaciones — candidatas a agrupación

**5. Interacción GrLivArea × Neighborhood**

| Extremo            | Valor                                          |
|:-------------------|:-----------------------------------------------|
| Correlacion minima | 0.32 (ClearCr)                                 |
| Correlacion maxima | 0.89 (NoRidge)                                 |
| Rango              | 0.57 — Hipotesis: crear feature de interaccion |

Rango de correlación GrLivArea-SalePrice según barrio

**6. No-linealidad OverallQual → SalePrice**

| Calidad | incremento\_\$ | flag         |
|--------:|---------------:|:-------------|
|       2 |         +1,620 |              |
|       3 |        +35,703 |              |
|       4 |        +20,947 |              |
|       5 |        +25,103 |              |
|       6 |        +28,080 |              |
|       7 |        +46,113 |              |
|       8 |        +67,019 |              |
|       9 |        +92,777 | salto grande |
|      10 |       +104,352 | salto grande |

Incremento medio de SalePrice por nivel de OverallQual

**7. Outliers extremos**

|                        | n_casos |
|:-----------------------|--------:|
| SalePrice \> 500,000   |       9 |
| LotArea \> 50,000 sqft |      10 |

Número de casos extremos detectados en el dataset

**8. Valores nulos (post-imputación)**

Variables criticas: 0% nulos \[OK\]

Esta documentación se genera automáticamente con los datos reales y te
ahorrará horas cuando vuelvas al proyecto semanas después o cuando
presentes resultados a tu equipo.

Añade celdas de Markdown entre bloques de código para explicar QUÉ estás
buscando y POR QUÉ es relevante. Un notebook sin narrativa es solo
código difícil de seguir.

### Preparación para modelado: *checklist* final

El *checklist* anterior cubre el proceso del EDA. Este segundo verifica
que el resultado de ese proceso es apto para modelar: no es redundante,
es la **puerta de salida** del EDA hacia la UT8. Antes de entrenar el
primer modelo, confirma:

**Variables:**

-   [ ] Variables finales seleccionadas documentadas
-   [ ] Multicolinealidad resuelta (\|r\| \< 0.8 entre predictores)
-   [ ] Categorías raras agrupadas o eliminadas

**Transformaciones:**

-   [ ] Variables asimétricas transformadas (log, sqrt, etc.)
-   [ ] Transformaciones documentadas (para invertir predicciones)
-   [ ] Escalado no necesario aún (se hace en UT8)

**Datos limpios:**

-   [ ] Nulos manejados (de UT5)
-   [ ] Outliers revisados y decisión tomada
-   [ ] Duplicados eliminados (de UT5)

**Documentación:**

-   [ ] Decisiones justificadas con EDA
-   [ ] Supuestos documentados
-   [ ] Baseline definido

### Errores comunes y cómo evitarlos

Has completado el EDA de Ames Housing: identificaste los 5 mejores
predictores, detectaste multicolinealidad entre
`GarageCars`/`GarageArea`, y confirmaste que `SalePrice` requiere
transformación logarítmica. ¿Qué tres decisiones concretas tomarías
antes de empezar a entrenar el modelo en UT8? Justifica cada una con un
hallazgo específico de este EDA.

## Del análisis a los datos: materializar los hallazgos del EDA

El informe de hallazgos que acabas de generar no es un resumen — es una
lista de decisiones pendientes. Cada sección del informe tiene una
traducción directa en el dataset:

| Hallazgo                                                                                                                         | Decisión                                                         |
|:----------------------------------------------|:------------------------|
| §1 Variables predictivas (r \> 0,35)                                                                                             | Conservar solo esas columnas                                     |
| §2 Skew \> 1: `SalePrice`, `LotArea`, `GrLivArea`                                                                                | Aplicar transformación log                                       |
| §3 Multicolinealidad: `GarageCars`/`GarageArea` (r=0.89); `GrLivArea`/`TotRmsAbvGrd` (r=0.83); `TotalBsmtSF`/`1stFlrSF` (r=0.80) | Eliminar uno de cada par redundante                              |
| §5 Interacción GrLivArea × Barrio detectada                                                                                      | BarrioPremium como proxy; GrLivArea_relativo se calculará en UT8 |

De los hallazgos del EDA a las decisiones de preparación del modelo

house_prices_ml.csv: 1458 filas × 21 cols
  · Columnas base seleccionadas: 16
  · Features derivadas del EDA: log_SalePrice, log_GrLivArea,
    log1p_TotalBsmtSF, BarrioPremium, PriceCategory
  · Eliminadas por multicolinealidad: GarageCars, TotRmsAbvGrd, 1stFlrSF, GarageYrBlt
  · Descartadas por baja correlación (r < 0,35): el resto

La interacción área-barrio (`GrLivArea_relativo`) se calculará en UT8
una vez separados los datos de entrenamiento y prueba — calcularla antes
introduciría *data leakage*.

## Fuentes y Lecturas Recomendadas

**¿Quieres profundizar más?** Consulta la bibliografía detallada, los
enlaces a la documentación oficial y los recursos de aprendizaje para
esta unidad en el **Apéndice B: Fuentes y Lecturas Recomendadas** al
final de este libro.

### Reflexión final